ingester-protocol/src/main/java/io/greptime/RouterClient.java

```diff
@@ -57,19 +63,41 @@ public boolean init(RouterOptions opts) {
         List<Endpoint> endpoints = Ensures.ensureNonNull(this.opts.getEndpoints(), "null `endpoints`");
 
         this.router = new DefaultRouter();
-        this.router.onRefresh(endpoints);
+        this.router.onRefresh(endpoints, null);
 
         long refreshPeriod = this.opts.getRefreshPeriodSeconds();
         if (refreshPeriod > 0) {
+            AtomicLong order = new AtomicLong(0);
             this.refresher = REFRESHER_POOL.getObject();
             this.refresher.scheduleWithFixedDelay(
-                    () -> this.router.refresh().whenComplete((r, e) -> {
-                        if (e != null) {
-                            LOG.error("Router cache refresh failed.", e);
-                        } else {
-                            LOG.debug("Router cache refresh {}.", r ? "success" : "failed");
-                        }
-                    }),
+                    () -> {
+                        long thisOrder = order.incrementAndGet();
+                        checkHealth().whenComplete((r, t) -> {
+                            if (t != null) {
+                                LOG.warn("Failed to check health", t);
+                                return;
+                            }
+
+                            // I don't want to worry about the overflow issue with long anymore,
+                            // because assuming one increment per second, it will take 292 years
+                            // to overflow. I think that's sufficient.
+                            if (thisOrder < order.get()) {
+                                LOG.warn("Skip outdated health check result, order: {}", thisOrder);
```

Comment:
I'm wondering if this "order" comparison is sufficient to guarantee the execution order of check health. Is the following executions possible:

a task is scheduled (task1):

thisOrder = 1 -> `checkHealth()` is called -> the "if" here is passed -> task1 is halted

a new task is scheduled (task2):

thisOrder = 2 -> `checkHealth()` is called -> the "if" here is passed -> the "refresh" below is executed -> task1 is resumed -> the "refresh" below is executed with outdated data, which is not intended.

Reply:
Good catch, fixed in https://github.com/GreptimeTeam/greptimedb-ingester-java/pull/50/commits/bc1ddebe5526aa5e0cee66ccad1f8d15cab63b2f